src/Utilities/model.jl

```diff
@@ -103,63 +103,107 @@ function MOI.add_variables(model::AbstractModel, n::Integer)
-        if f.variable == vi
-            push!(rm, ci)
+        if vi in f.variables
+            if length(f.variables) > 1
+                # If `supports_dimension_update(s)` then the variable will be
+                # removed in `_remove_variable`.
+                if !supports_dimension_update(typeof(s))
+                    throw_delete_variable_in_vov(vi)
+                end
+            else
+                push!(rm, ci)
+            end
         end
     end
     return rm
 end
-function MOI.delete(model::AbstractModel, vi::VI)
-    if !MOI.is_valid(model, vi)
-        throw(MOI.InvalidIndex(vi))
+function _vector_of_variables_with(
+    constrs::Vector{<:ConstraintEntry{MOI.VectorOfVariables}},
+    vis::Vector{VI}
+)
+    rm = CI{MOI.VectorOfVariables}[]
+    for (ci, f, s) in constrs
+        if vis == f.variables
+            push!(rm, ci)
+        end
     end
-    model.objective = removevariable(model.objective, vi)
-    # `ci_to_remove` is the list of indices of the `SingleVariable` constraints
-    # of `vi`
-    ci_to_remove = broadcastvcat(constrs -> _removevar!(constrs, vi), model)
-    for ci in ci_to_remove
+    return rm
+end
+function MOI.delete(model::AbstractModel{T}, vi::VI) where T
+    MOI.throw_if_not_valid(model, vi)
+    model.objective = remove_variable(model.objective, vi)
+    # If a variable is removed, the `VectorOfVariables` constraints using this
+    # variable only need to be removed too. `vov_to_remove` is the list of
+    # indices of the `VectorOfVariables` constraints of `vi`.
+    vov_to_remove = broadcastvcat(constrs -> _vector_of_variables_with(constrs, vi), model)
+    for ci in vov_to_remove
         MOI.delete(model, ci)
     end
+    # `VectorOfVariables` constraints with sets not supporting dimension update
+    # were either deleted or an error was thrown. The rest is modified now.
+    broadcastcall(constrs -> _remove_variable(constrs, vi), model)
     model.single_variable_mask[vi.value] = 0x0
     if model.variable_indices === nothing
         model.variable_indices = Set(MOI.get(model,
                                              MOI.ListOfVariableIndices()))
     end
     delete!(model.variable_indices, vi)
     model.name_to_var = nothing
-    if haskey(model.var_to_name, vi)
-        delete!(model.var_to_name, vi)
+    delete!(model.var_to_name, vi)
+    model.name_to_con = nothing
+    delete!(model.con_to_name, MOI.ConstraintIndex{MOI.SingleVariable, MOI.EqualTo{T}}(vi.value))
+    delete!(model.con_to_name, MOI.ConstraintIndex{MOI.SingleVariable, MOI.GreaterThan{T}}(vi.value))
+    delete!(model.con_to_name, MOI.ConstraintIndex{MOI.SingleVariable, MOI.LessThan{T}}(vi.value))
+    delete!(model.con_to_name, MOI.ConstraintIndex{MOI.SingleVariable, MOI.Interval{T}}(vi.value))
+    delete!(model.con_to_name, MOI.ConstraintIndex{MOI.SingleVariable, MOI.Integer}(vi.value))
+    delete!(model.con_to_name, MOI.ConstraintIndex{MOI.SingleVariable, MOI.ZeroOne}(vi.value))
```

Comment:
This list is missing `Semicontinuous` and `Semiinteger`. This list is also dangerous because if you add another type of set to the list of sets handled by `single_variable_flag`, it's easy to forget updating this list. At least, add a comment on `single_variable_flag` saying that if the list of sets is modified, then you should also modify the list here.